File: static/js/mentor.js
// The mentor flow: level-up verification and the proposal inbox.
//
// A level-up is never a button that just increments a number. The order is
// deliberate and strict: notes first (no writeup, no attempt), then the bar
// is shown up front, then the mentor asks questions grounded in the notes,
// and only answers that hold up get the level. "Not yet" is a real verdict.

async function startLevelUp(nodeId, opts = {}) {
  closeModal();
  openModal(`<div class="mentor-wait">
    <div class="mentor-badge">MENTOR</div>
    <p>Checking the bar…</p>
  </div>`);

  let preview, docs;
  try {
    [preview, docs] = await Promise.all([
      API.get(`/tree/nodes/${nodeId}/levelup/preview`),
      API.get("/docs"),
    ]);
  } catch (e) {
    closeModal();
    return;
  }

  const diff = preview.difficulty;
  const fails = preview.rejected_attempts;

  openModal(`
    <div class="modal-head">
      <div>
        <div class="mentor-badge">MENTOR</div>
        <h2 class="modal-title">Level ${preview.target_level} — the bar</h2>
      </div>
      <button class="icon-btn" onclick="closeModal()">×</button>
    </div>

    <div class="difficulty-row">
      <span class="card-meta">Difficulty</span>
      ${Array.from({ length: 5 }, (_, i) =>
        `<span class="pip ${i < diff ? "on" : ""}"></span>`).join("")}
      <span class="card-meta">${diff}/5</span>
      ${fails ? `<span class="chip static c-red">${fails} failed attempt${fails > 1 ? "s" : ""}</span>` : ""}
    </div>

    <div class="expectation-box">
      <div class="expectation-title">Expect to prove</div>
      <p>${esc(preview.expectations)}</p>
    </div>

    <label class="field-label">Your notes (required)</label>
    <select id="evidence-doc">
      <option value="">— pick a writeup —</option>
      ${docs.map((d) => `<option value="${d.id}">${esc(d.title)}</option>`).join("")}
    </select>
    <p class="notes-gate-hint">
      Notes come first. A real writeup — what you did, the commands or
      techniques, and <em>why</em> they worked — at least
      ${preview.notes_min_chars} characters. Thin notes are rejected before
      the mentor even asks a question.
    </p>

    <div class="modal-actions">
      <button class="btn" id="write-notes">Write notes first</button>
      <button class="btn primary" id="begin-attempt">Begin verification</button>
    </div>
  `, () => {
    el("write-notes").addEventListener("click", () => {
      closeModal();
      go("docs");
      toast("Write the room/skill up, then come back to the node", "info", 5000);
    });

    el("begin-attempt").addEventListener("click", async () => {
      const evidence = el("evidence-doc").value;
      if (!evidence) {
        toast("Pick your notes doc — no writeup, no attempt", "error");
        return;
      }
      openModal(`<div class="mentor-wait">
        <div class="mentor-badge">MENTOR</div>
        <p>Reading your notes, preparing questions…</p>
      </div>`);

      let attempt;
      try {
        attempt = await API.post(`/tree/nodes/${nodeId}/levelup`, {
          evidence_doc: Number(evidence),
          room_code: opts.roomCode || null,
        });
      } catch (e) {
        // The API client already toasted the notes-gate message.
        startLevelUp(nodeId, opts);
        return;
      }

      const questions = JSON.parse(attempt.questions || "[]");
      if (!questions.length) {
        showQueued(attempt);
        return;
      }
      showQuestions(attempt.id, attempt, questions);
    });
  });
}

function showQueued(attempt) {
  openModal(`
    <div class="modal-head">
      <h2 class="modal-title">Verification queued</h2>
      <button class="icon-btn" onclick="closeModal()">×</button>
    </div>
    <p class="node-desc">
      Attempt <strong>#${attempt.id}</strong> is open at difficulty
      ${attempt.difficulty}/5, notes attached, waiting on questions.
    </p>
    <p class="node-desc">
      Set <code>ANTHROPIC_API_KEY</code> to have the mentor respond in-app,
      or have Claude Code pick it up:
    </p>
    <pre class="code-hint">GET  /api/attempts?status=awaiting_questions
POST /api/attempts/${attempt.id}/questions</pre>
    <div class="modal-actions">
      <button class="btn danger" id="cancel-attempt">Cancel attempt</button>
      <button class="btn" onclick="closeModal()">Close</button>
    </div>
  `, () => {
    el("cancel-attempt").addEventListener("click", async () => {
      await API.del(`/attempts/${attempt.id}`);
      closeModal();
      renderTree();
    });
  });
}

async function openAttempt(attemptId) {
  closeModal();
  const attempt = await API.get(`/attempts/${attemptId}`);
  const questions = attempt.questions || [];
  if (!questions.length) {
    toast("Still waiting on the mentor's questions", "info");
    return;
  }
  showQuestions(attemptId, attempt, questions, attempt.answers || []);
}

async function showQuestions(attemptId, attempt, questions, existing = []) {
  const title = attempt.node_title || attempt.context?.node?.title || "this skill";
  const diff = attempt.difficulty || attempt.context?.difficulty || 1;

  let notesTitle = "";
  if (attempt.evidence_doc) {
    try { notesTitle = (await API.get(`/docs/${attempt.evidence_doc}`)).title; }
    catch (e) { /* doc deleted mid-attempt; the server still has nothing to grade against */ }
  }

  openModal(`
    <div class="modal-head">
      <div>
        <div class="mentor-badge">MENTOR</div>
        <h2 class="modal-title">${esc(title)}</h2>
      </div>
      <button class="icon-btn" onclick="closeModal()">×</button>
    </div>

    <div class="difficulty-row">
      <span class="card-meta">Difficulty</span>
      ${Array.from({ length: 5 }, (_, i) =>
        `<span class="pip ${i < diff ? "on" : ""}"></span>`).join("")}
      <span class="card-meta">${diff}/5</span>
      ${notesTitle ? `<span class="chip static c-teal">notes: ${esc(truncate(notesTitle, 24))}</span>` : ""}
    </div>

    <p class="mentor-intro">
      Answer in your own words. The mentor has read your notes and will
      grade both together — specifics beat definitions, and vague answers
      are rejected, not nudged.
    </p>

    ${questions.map((q, i) => `
      <div class="question-block">
        <div class="question-text"><span class="qnum">${i + 1}</span>${esc(q)}</div>
        <textarea class="modal-textarea" data-answer="${i}" rows="3"
                  placeholder="Your answer…">${esc(existing[i] || "")}</textarea>
      </div>`).join("")}

    <div class="modal-actions">
      <button class="btn" id="not-ready" title="Closes the attempt without a verdict — it doesn't count as a failure">
        Not ready yet
      </button>
      <button class="btn primary" id="submit-answers">Submit for review</button>
    </div>
  `, (modal) => {
    el("not-ready").addEventListener("click", async () => {
      await API.del(`/attempts/${attemptId}`);
      closeModal();
      toast("Attempt closed — no verdict, no failure. Come back when you're ready.", "info", 4500);
      renderTree();
    });

    el("submit-answers").addEventListener("click", async () => {
      const answers = [...modal.querySelectorAll("[data-answer]")].map((t) => t.value.trim());
      if (answers.every((a) => !a)) {
        toast("Answer at least one question", "error");
        return;
      }

      openModal(`<div class="mentor-wait">
        <div class="mentor-badge">MENTOR</div>
        <p>Reviewing your answers against your notes…</p>
      </div>`);

      let result;
      try {
        result = await API.post(`/attempts/${attemptId}/answer`, { answers });
      } catch (e) {
        closeModal();
        return;
      }

      if (result.status === "grading") {
        openModal(`
          <div class="modal-head">
            <h2 class="modal-title">Submitted</h2>
            <button class="icon-btn" onclick="closeModal()">×</button>
          </div>
          <p class="node-desc">Your answers are queued for review.</p>
          <pre class="code-hint">GET  /api/attempts?status=grading
POST /api/attempts/${attemptId}/verdict</pre>
          <div class="modal-actions">
            <button class="btn" onclick="closeModal()">Close</button>
          </div>`);
        return;
      }

      showVerdict(result);
    });
  });
}

function showVerdict(result) {
  const granted = result.granted;

  const unlocks = (result.unlocked_nodes || []).length
    ? `<div class="unlock-note">
         Unlocked: ${result.unlocked_nodes.map((n) => esc(n.title)).join(", ")}
       </div>`
    : "";

  const proposal = result.proposal_id
    ? `<div class="unlock-note subtle">
         The mentor suggested new nodes — waiting in Growth › proposals.
       </div>`
    : "";

  openModal(`
    <div class="modal-head">
      <div>
        <div class="mentor-badge ${granted ? "granted" : "rejected"}">
          ${granted ? "LEVEL GRANTED" : "NOT YET"}
        </div>
        ${granted ? `<h2 class="modal-title">Level ${result.new_level}</h2>` : ""}
      </div>
      <button class="icon-btn" onclick="closeModal()">×</button>
    </div>

    <p class="verdict-feedback">${esc(result.feedback)}</p>
    ${granted ? "" : `<p class="mentor-intro">
      A failed attempt stays on the record. Close the gap it named — in your
      notes and in your head — and try again.
    </p>`}
    ${unlocks}
    ${proposal}

    <div class="modal-actions">
      ${granted
        ? `<button class="btn primary" id="verdict-close">Continue</button>`
        : `<button class="btn" id="verdict-close">Close</button>
           <button class="btn primary" id="retry">Try again</button>`}
    </div>
  `, () => {
    el("verdict-close").addEventListener("click", async () => {
      closeModal();
      await renderTree();
      if (granted) celebrateLevelUp(result.node_id);
    });
    el("retry")?.addEventListener("click", async () => {
      closeModal();
      await renderTree();
      startLevelUp(result.node_id);
    });
  });
}

// ---------- Proposals ----------
async function renderProposalsInto(container) {
  const proposals = await API.get("/proposals?status=pending");

  if (!proposals.length) {
    container.innerHTML = `<p class="empty-state">No pending proposals.</p>`;
    return;
  }

  container.innerHTML = proposals.map((p) => `
    <div class="proposal" data-proposal="${p.id}">
      <div class="proposal-head">
        <span class="chip static c-purple">${esc(p.kind.replace("_", " "))}</span>
        <span class="proposal-title">${esc(p.title)}</span>
      </div>
      ${p.rationale ? `<p class="proposal-rationale">${esc(p.rationale)}</p>` : ""}
      <div class="proposal-actions-list">
        ${p.actions.map((a) => `<div class="action-line">${esc(describeAction(a))}</div>`).join("")}
      </div>
      <div class="proposal-buttons">
        <button class="btn" data-reject="${p.id}">Reject</button>
        <button class="btn primary" data-approve="${p.id}">Approve</button>
      </div>
    </div>`).join("");

  container.querySelectorAll("[data-approve]").forEach((btn) => {
    btn.addEventListener("click", async () => {
      const res = await API.post(`/proposals/${btn.dataset.approve}/approve`, {});
      toast(res.errors?.length
        ? `Applied with ${res.errors.length} error(s)`
        : `Applied ${res.applied.length} change(s)`,
        res.errors?.length ? "error" : "info");
      renderGrowth();
    });
  });

  container.querySelectorAll("[data-reject]").forEach((btn) => {
    btn.addEventListener("click", async () => {
      await API.post(`/proposals/${btn.dataset.reject}/reject`, {});
      renderGrowth();
    });
  });
}

// Turn an action object into a line a human can actually check before approving.
function describeAction(a) {
  switch (a.op) {
    case "create_node":  return `Add node "${a.title}" (${a.domain}, tier ${a.tier})`;
    case "update_node":  return `Edit node #${a.node_id}`;
    case "create_edge":  return `Connect node #${a.from_id} → #${a.to_id}`;
    case "delete_edge":  return `Disconnect #${a.from_id} → #${a.to_id}`;
    case "move_card":    return `Move card #${a.card_id} to list #${a.list_id}`;
    case "set_due":      return `Set card #${a.card_id} due ${a.due_at || "(cleared)"}`;
    case "update_card":  return `Update card #${a.card_id}`;
    case "create_card":  return `Create card "${a.title}"`;
    case "create_routine": return `Add routine "${a.name}"`;
    default:             return a.op || "unknown action";
  }
}

File: mentor.py
"""
The mentor: verification of skill level-ups, and applying AI proposals.

Two modes, both using the same data:

  Queue mode (always on). A level-up attempt is a real row moving through
  states, and an external agent (Claude Code, a cron script, you with curl)
  reads pending attempts, posts questions, then posts a verdict. Nothing is
  hidden - the whole handshake is in the API.

  Direct mode (optional). If ANTHROPIC_API_KEY is set, the server calls the
  Anthropic API itself so verification works without a second tool open.
  Direct mode is a convenience layer on top of the queue, not a bypass: it
  writes the same rows, so you can start an attempt in one mode and finish
  it in the other.

The mentor's job is to be a challenger, not a rubber stamp. Difficulty
scales with tree depth AND existing attribute strength, so verification
gets harder as you get better rather than flattening out.
"""
import json
import os
import urllib.error
import urllib.request

API_KEY = os.environ.get("ANTHROPIC_API_KEY", "")
MODEL = os.environ.get("OPSDECK_MENTOR_MODEL", "claude-sonnet-4-6")
ANTHROPIC_URL = "https://api.anthropic.com/v1/messages"

DIFFICULTY_GUIDE = {
    1: "foundational. One or two concrete questions. Confirm they actually "
       "understand the basics rather than having read about them.",
    2: "working knowledge. Ask them to explain a mechanism, not recite a "
       "definition. At least one question should require an example.",
    3: "practitioner. Require specifics from real work: a command they ran, "
       "output they interpreted, a decision they made and why.",
    4: "advanced. Probe edge cases and failure modes. Ask what breaks, what "
       "they'd check first, what a naive approach gets wrong.",
    5: "expert. Demand depth a textbook answer cannot fake - tradeoffs, "
       "things they've personally debugged, where the common advice is wrong.",
}

MENTOR_SYSTEM = """You are the mentor in a cybersecurity student's skill-tracking app.

You are a strict examiner, not an assistant and not a cheerleader. This \
person is trying to become genuinely excellent, and they have explicitly \
asked you to hold the bar and gatekeep. They chose this. Every level you \
grant is a claim about reality, and your default posture is skepticism: \
the burden of proof is on them, always.

Rules:
- Ask real, specific questions about the actual skill. Never generic quiz \
questions, never anything answerable from a definition page. Prefer \
questions grounded in their own notes - test whether they understand what \
they wrote, not just that they wrote something.
- Scale difficulty to what you are told. A tier-1 node with a beginner is a \
different conversation than a tier-4 node with someone already strong there. \
Advanced levels require demonstrated depth: a real scenario walked through, \
a tradeoff explained, a decision defended.
- Vague, hand-wavy, or terminology-reciting answers are rejected outright. \
Not softened, not "could be stronger" - rejected, with a plain statement of \
exactly what was missing and what would convince you.
- Thin notes are rejected the same as thin answers. If the attached writeup \
is a one-liner, a pasted walkthrough with nothing of their own, or steps \
with no reasoning about *why* anything worked, say so and refuse.
- "No, not yet" is a legitimate final verdict. You are not tuned so that \
attempts eventually succeed. Repeated failure on the same node is an \
acceptable outcome; passing someone who hasn't earned it is not.
- When an answer genuinely clears the bar, grant it plainly and say what \
was strong. Strictness is about the rigor of the check, not manufactured \
friction - do not reject good work to seem tough.
- Be direct and unsparing about the verdict without being demeaning about \
the person. Strict grader, not a bully. Do not flatter, pad, or apologize.

You are not a customer service agent. You are the person who tells them the \
truth about where they actually stand."""


def available():
    return bool(API_KEY)


def _call(system, user, max_tokens=1200):
    """Single Anthropic API call returning concatenated text blocks."""
    payload = {
        "model": MODEL,
        "max_tokens": max_tokens,
        "system": system,
        "messages": [{"role": "user", "content": user}],
    }
    req = urllib.request.Request(
        ANTHROPIC_URL,
        data=json.dumps(payload).encode(),
        headers={
            "content-type": "application/json",
            "x-api-key": API_KEY,
            "anthropic-version": "2023-06-01",
        },
        method="POST",
    )
    with urllib.request.urlopen(req, timeout=90) as res:
        data = json.loads(res.read().decode())
    return "".join(b.get("text", "") for b in data.get("content", []) if b.get("type") == "text")


def _extract_json(text):
    """Pull a JSON object out of a reply that may be fenced or prefaced."""
    text = text.strip()
    if text.startswith("```"):
        text = text.split("```")[1]
        if text.startswith("json"):
            text = text[4:]
    start, end = text.find("{"), text.rfind("}")
    if start == -1 or end == -1:
        raise ValueError("no JSON object in reply")
    return json.loads(text[start:end + 1])


def generate_questions(context, notes=None):
    """
    Ask the mentor for verification questions, grounded in the user's own
    notes when provided. Returns a list of question strings. Raises if
    direct mode is unavailable.
    """
    if not available():
        raise RuntimeError("ANTHROPIC_API_KEY not set - use queue mode instead")

    d = context["difficulty"]
    notes_block = (
        f"\n\nTheir notes for this attempt (mandatory reading - base at least "
        f"one question directly on something they wrote):\n{notes[:6000]}"
        if notes else ""
    )
    prompt = f"""A skill level-up has been requested. Write the verification questions.

Context:
{json.dumps(context, indent=2)}

Difficulty {d}/5 - {DIFFICULTY_GUIDE.get(d, DIFFICULTY_GUIDE[3])}{notes_block}

Write {2 if d <= 2 else 3} questions specific to "{context['node']['title']}". \
They must be answerable only by someone who has actually done this, not \
someone who has read about it. If notes were provided, probe the weakest or \
least-explained part of them - the goal is to test whether they understand \
what they wrote.

Respond with JSON only, no other text:
{{"questions": ["...", "..."]}}"""

    return _extract_json(_call(MENTOR_SYSTEM, prompt))["questions"]


def grade_answers(context, questions, answers, evidence=None):
    """
    Ask the mentor to judge the answers.
    Returns {"granted": bool, "feedback": str, "suggested_nodes": [...]}.
    """
    if not available():
        raise RuntimeError("ANTHROPIC_API_KEY not set - use queue mode instead")

    qa = "\n\n".join(
        f"Q{i+1}: {q}\nA{i+1}: {a or '(no answer given)'}"
        for i, (q, a) in enumerate(zip(questions, answers))
    )
    ev = (f"\n\nTheir notes for this attempt:\n{evidence[:6000]}"
          if evidence else "\n\n(No notes were attached.)")
    d = context["difficulty"]

    prompt = f"""Judge whether this level-up is earned. Default to no; they must convince you.

Context:
{json.dumps(context, indent=2)}

Difficulty {d}/5 - {DIFFICULTY_GUIDE.get(d, DIFFICULTY_GUIDE[3])}

{qa}{ev}

Judge the notes AND the answers together. Grant only if both hold up: the \
notes show real work of their own (specific commands, techniques, reasoning \
about why things worked - not a pasted walkthrough) and the answers \
demonstrate they understand what they wrote. Reject for: vague answers, \
recited terminology, dodged specifics, or notes too thin to defend. When \
rejecting, state exactly what was missing and what would convince you next \
time - "you documented the steps but not why the exploit worked" beats \
"needs more detail".

You may also suggest 1-2 adjacent skill nodes worth adding to their tree \
based on what they demonstrated. Only suggest nodes that follow naturally \
from this one.

Respond with JSON only, no other text:
{{"granted": true/false,
  "feedback": "2-4 sentences, direct, addressed to them",
  "suggested_nodes": [{{"title": "...", "domain": "...", "tier": 2,
                        "rationale": "...",
                        "weights": [{{"attribute_key": "pentest", "weight": 1.0}}]}}]}}"""

    out = _extract_json(_call(MENTOR_SYSTEM, prompt, max_tokens=1600))
    out.setdefault("suggested_nodes", [])
    out.setdefault("feedback", "")
    out["granted"] = bool(out.get("granted"))
    return out


def recommend_room(context):
    """
    Ask the mentor to point the user at their next TryHackMe room(s), based
    on weakest/stalest attributes and what's already been completed.
    Returns {"recommendations": [{room_code, room_title, reason, node_ids}], "summary": str}.
    """
    if not available():
        raise RuntimeError("ANTHROPIC_API_KEY not set - use queue mode instead")

    prompt = f"""Recommend the user's next TryHackMe work.

Their current state:
{json.dumps(context, indent=2)}

Rules:
- Target the weakest or stalest attributes first, using tree node levels to
  find the specific sub-area that's lagging (e.g. "Pentest is fine overall
  but privilege escalation on Windows has no levels").
- Recommend 1-3 specific, real TryHackMe rooms by their exact room code (the
  URL slug) and title. Never recommend anything in completed_rooms.
- If you are not certain a room code is exactly right, say so in the reason
  and give the room title so the user can search for it - do not invent
  plausible-looking codes silently.
- Each recommendation names which existing tree node(s) it should feed
  (node_ids from tree_nodes), so a completion can trigger verification there.
- Be specific about WHY: "you're weak on X relative to Y, this room covers
  exactly that" - not a generic list.

Respond with JSON only, no other text:
{{"summary": "1-2 sentences on where their gaps are",
  "recommendations": [{{"room_code": "...", "room_title": "...",
                        "reason": "...", "node_ids": [1, 2]}}]}}"""

    out = _extract_json(_call(MENTOR_SYSTEM, prompt, max_tokens=1400))
    out.setdefault("recommendations", [])
    out.setdefault("summary", "")
    return out


# ------------------------------------------------- applying proposals

def apply_proposal(conn, proposal):
    """
    Execute an approved proposal's action list.

    Actions are deliberately a small vocabulary rather than arbitrary SQL -
    an approved proposal should not be able to do anything the API couldn't,
    and the user approving a summary shouldn't be authorising a blank cheque.
    """
    actions = json.loads(proposal["actions"] or "[]")
    applied, errors = [], []

    for a in actions:
        op = a.get("op")
        try:
            if op == "move_card":
                conn.execute(
                    "UPDATE cards SET list_id=?, position=? WHERE id=?",
                    (a["list_id"], a.get("position", 0), a["card_id"]),
                )
            elif op == "set_due":
                conn.execute(
                    "UPDATE cards SET due_at=? WHERE id=?", (a.get("due_at"), a["card_id"])
                )
            elif op == "update_card":
                for f in ("title", "description", "completed", "archived"):
                    if f in a:
                        conn.execute(f"UPDATE cards SET {f}=? WHERE id=?", (a[f], a["card_id"]))
            elif op == "create_card":
                pos = conn.execute(
                    "SELECT COALESCE(MAX(position),-1)+1 FROM cards WHERE list_id=?",
                    (a["list_id"],),
                ).fetchone()[0]
                conn.execute(
                    "INSERT INTO cards (list_id,title,description,due_at,position) VALUES (?,?,?,?,?)",
                    (a["list_id"], a.get("title", "Untitled"), a.get("description", ""),
                     a.get("due_at"), pos),
                )
            elif op == "create_node":
                owner = a.get("profile_id") or (
                    proposal["profile_id"] if "profile_id" in proposal.keys() else "primary")
                cur = conn.execute(
                    """INSERT INTO skill_nodes (title,description,domain,x,y,tier,max_level,
                                                unlock_attr,unlock_value,profile_id)
                       VALUES (?,?,?,?,?,?,?,?,?,?)""",
                    (a["title"], a.get("description", ""), a.get("domain", "general"),
                     a.get("x", 0), a.get("y", 0), a.get("tier", 1), a.get("max_level", 5),
                     a.get("unlock_attr"), a.get("unlock_value"), owner),
                )
                nid = cur.lastrowid
                for w in a.get("weights", []):
                    conn.execute(
                        "INSERT OR REPLACE INTO node_weights (node_id,attribute_key,weight) VALUES (?,?,?)",
                        (nid, w["attribute_key"], w.get("weight", 1.0)),
                    )
                for parent in a.get("parents", []):
                    conn.execute(
                        "INSERT OR IGNORE INTO skill_edges (from_id,to_id) VALUES (?,?)",
                        (parent, nid),
                    )
            elif op == "update_node":
                for f in ("title", "description", "domain", "x", "y", "tier",
                          "max_level", "unlock_attr", "unlock_value"):
                    if f in a:
                        conn.execute(
                            f"UPDATE skill_nodes SET {f}=? WHERE id=?", (a[f], a["node_id"])
                        )
            elif op == "create_edge":
                conn.execute(
                    "INSERT OR IGNORE INTO skill_edges (from_id,to_id) VALUES (?,?)",
                    (a["from_id"], a["to_id"]),
                )
            elif op == "delete_edge":
                conn.execute(
                    "DELETE FROM skill_edges WHERE from_id=? AND to_id=?",
                    (a["from_id"], a["to_id"]),
                )
            elif op == "create_routine":
                conn.execute(
                    "INSERT INTO routines (name,time_group,notes,profile_id) VALUES (?,?,?,?)",
                    (a["name"], a.get("time_group", "anytime"), a.get("notes", ""),
                     a.get("profile_id", proposal["profile_id"] if "profile_id" in proposal.keys() else "primary")),
                )

            # ---- destructive ops -------------------------------------
            # These delete rows for real. They still route through the
            # proposal queue, so nothing here runs until the user has read a
            # plain-language summary and approved it. FK cascades mean
            # deleting a parent takes its children (a board takes its lists
            # and cards) - that is intended, and is why the summary should
            # say so.
            elif op == "delete_card":
                conn.execute("DELETE FROM cards WHERE id=?", (a["card_id"],))
            elif op == "delete_list":
                conn.execute("DELETE FROM lists WHERE id=?", (a["list_id"],))
            elif op == "delete_board":
                conn.execute("DELETE FROM boards WHERE id=?", (a["board_id"],))
            elif op == "delete_node":
                conn.execute("DELETE FROM skill_nodes WHERE id=?", (a["node_id"],))
            elif op == "delete_routine":
                conn.execute("DELETE FROM routines WHERE id=?", (a["routine_id"],))
            elif op == "delete_doc":
                conn.execute("DELETE FROM docs WHERE id=?", (a["doc_id"],))
            elif op == "delete_event":
                conn.execute("DELETE FROM events WHERE id=?", (a["event_id"],))
            elif op == "delete_label":
                conn.execute("DELETE FROM labels WHERE id=?", (a["label_id"],))
            elif op == "delete_checklist_item":
                conn.execute("DELETE FROM checklist_items WHERE id=?", (a["item_id"],))
            elif op == "delete_attribute":
                conn.execute("DELETE FROM attributes WHERE id=?", (a["attribute_id"],))

            else:
                errors.append(f"unknown op: {op}")
                continue
            applied.append(op)
        except Exception as exc:  # keep going; report what failed
            errors.append(f"{op}: {exc}")

    return {"applied": applied, "errors": errors}
